src/library/scala/collection/immutable/MapBuilder.scala

```diff
@@ -0,0 +1,39 @@
+package scala.collection.immutable
+
+import scala.collection.mutable.ReusableBuilder
+import scala.reflect.{ClassTag, classTag}
+/**
+  * A builder specialised for immutable Maps
+  *
+  * Note - this builder can be serially reused - it is NOT required to call `clear` after `result`
+  */
+class MapBuilder [K, V, C <: Map[K, V] : ClassTag ] private[immutable](empty: C)
+  extends ReusableBuilder[(K,V), C] {
+
+  protected var elems: C = empty
+
+  def clear(): Unit = { elems = empty }
+
+  def result(): C = elems
+
+  final def size: Int = elems.size
+
+  final def isEmpty: Boolean = elems.isEmpty
+
+  /**
+    * are the types implementationally compatible
+    */
+  protected def isCompatibleType(value: IterableOnce[(K, V)]): Boolean =
+    classTag[C].runtimeClass.isInstance(value)
+
+  def addOne(elem: (K, V)): this.type = {
+    elems = (elems + elem).asInstanceOf[C]; this
+  }
```

Comment:
Should we have `def addKeyValue(k: K, v: V)` here? to allow construction without tupling?

Reply:
done